src/omp/exec_dist.f90

```diff
@@ -61,5 +61,134 @@ subroutine exec_dist_tds_compact( &
+      real(dp), dimension(:, :, :), intent(out) :: &
+         d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e
+
+      real(dp), dimension(:, :, :), intent(in) :: u, u_recv_s, u_recv_e
+      real(dp), dimension(:, :, :), intent(in) :: v, v_recv_s, v_recv_e
+
+      type(tdsops_t), intent(in) :: tdsops_du, tdsops_dud, tdsops_d2u
+
+      real(dp), dimension(:, :), allocatable :: ud, ud_recv_s, ud_recv_e
+      real(dp) :: nu
+      integer, intent(in) :: nproc, pprev, pnext
+      integer, intent(in) :: n_block
+
+      integer :: n_data, n_halo
+      integer :: k, i, j, n
+
+      ! TODO: don't hardcode n_halo
+      n_halo = 4
+      n = tdsops_d2u%n
+      n_data = SZ*n_block
+
+      allocate(ud(SZ, n))
+      allocate(ud_recv_e(SZ, n_halo))
+      allocate(ud_recv_s(SZ, n_halo))
+
+      !$omp parallel do
+      do k = 1, n_block
+         call der_univ_dist( &
+            du(:, :, k), du_send_s(:, :, k), du_send_e(:, :, k), u(:, :, k), &
+            u_recv_s(:, :, k), u_recv_e(:, :, k), &
+            tdsops_du%coeffs_s, tdsops_du%coeffs_e, tdsops_du%coeffs, tdsops_du%n, &
+            tdsops_du%dist_fw, tdsops_du%dist_bw, tdsops_du%dist_af &
+            )
+
+         call der_univ_dist( &
+            d2u(:, :, k), d2u_send_s(:, :, k), d2u_send_e(:, :, k), u(:, :, k), &
+            u_recv_s(:, :, k), u_recv_e(:, :, k), &
+            tdsops_d2u%coeffs_s, tdsops_d2u%coeffs_e, tdsops_d2u%coeffs, tdsops_d2u%n, &
+            tdsops_d2u%dist_fw, tdsops_d2u%dist_bw, tdsops_d2u%dist_af &
+            )
+
+         ! Handle dud by locally generating u*v
+         do j = 1, n
+            !$omp simd
+            do i = 1, SZ
+               ud(i, j) = u(i, j, k) * v(i, j, k)
+            end do
+            !$omp end simd
+         end do
+
+         do j = 1, n_halo
+            !$omp simd
+            do i = 1, SZ
+               ud_recv_s(i, j) = u_recv_s(i, j, k) * v_recv_s(i, j, k)
+               ud_recv_e(i, j) = u_recv_e(i, j, k) * v_recv_e(i, j, k)
+            end do
+            !$omp end simd
+         end do
+
+         call der_univ_dist( &
+            dud(:, :, k), dud_send_s(:, :, k), dud_send_e(:, :, k), ud(:, :), &
+            ud_recv_s(:, :), ud_recv_e(:, :), &
+            tdsops_dud%coeffs_s, tdsops_dud%coeffs_e, tdsops_dud%coeffs, tdsops_dud%n, &
+            tdsops_dud%dist_fw, tdsops_dud%dist_bw, tdsops_dud%dist_af &
+            )
+       
+      end do
+      !$omp end parallel do
+
+      ! halo exchange for 2x2 systems
+      call sendrecv_fields(du_recv_s, du_recv_e, du_send_s, du_send_e, &
+                           n_data, nproc, pprev, pnext)
+      call sendrecv_fields(dud_recv_s, dud_recv_e, dud_send_s, dud_send_e, &
+                           n_data, nproc, pprev, pnext)
+      call sendrecv_fields(d2u_recv_s, d2u_recv_e, d2u_send_s, d2u_send_e, &
+                           n_data, nproc, pprev, pnext)
+
+      !$omp parallel do
+      do k = 1, n_block
+         call der_univ_subs(du(:, :, k), &
```

Comment:
I realised that here we're writing 3 field sized arrays into main memory unnecessarily. It is potentially increasing the runtime %20.

In the second phase of the algorithm here we pass a part of the `du`, `dud`, and `d2u` into der_univ_subs, and they're all rewritten in place. Then later we combine them in `rhs` for the final result. Ideally, we want `du`, `dud`, and `d2u` to be read once and `rhs` to be written only once. However because of the way der_univ_subs work, the updated data in `du` arrays after der_univ_subs call gets written in the main memory, even though we don't need this at all.

There are three ways we can fix this
- In the parallel do loop in the second phase we can copy the relevant parts of `du`, `dud`, and `d2u` arrays into `(SZ, n)` sized temporary arrays. Then we pass temporary arrays into der_univ_subs, and at the end we use these temporaries to obtain final `rhs`. This is the easiest solution but it may not be the best in terms of performance.
- We can write an alternative der_univ_subs and separate input and output arrays. This way we can pass a part of the `du` arrays as we do now, and pass a small temporary array as the output one. Because `du` arrays will be input arrays no data will be written in main memory. Then we can combine the temporaries to get `rhs`.
- If we're writing an alternative der_univ_subs to be used in transeq, we can go one step further and have a fused version of it. This would probably the most performant solution. der_univ_subs is relatively lightweight so it isn't really hard to do so. The new subrotuine can input all `du`, `dud`, and `d2u`, and write the final result `rhs`.

Reply:
#40 is relevant here

Reply:
Good points, I will have a think about it, but indeed having it in a new PR focusing on performance makes sense.

Reply:
Alright, I think best way to move forward is implementing the first strategy and checking how much of the peak BW we get. If its a good utilisation then maybe we don't need to bother with a new der_univ_subs at all.